driver-core/src/main/com/mongodb/internal/connection/LoadBalancedServer.java

```diff
@@ -0,0 +1,163 @@
+                ServerDescription.builder()
+                        .ok(true)
+                        .state(ServerConnectionState.CONNECTED)
+                        .type(ServerType.LOAD_BALANCER)
+                        .address(serverId.getAddress())
+                        .build(),
+                ServerDescription.builder().address(serverId.getAddress()).state(CONNECTING).build()));
+    }
+
+    @Override
+    public void resetToConnecting() {
+        // no op
+    }
+
+    @Override
+    public void invalidate() {
+        // no op
+    }
+
+    @Override
+    public void invalidate(final ConnectionState connectionState, final Throwable reason, final int connectionGeneration,
+                           final int maxWireVersion) {
+        invalidate(connectionState, reason, null);
+    }
+
+
+    private void invalidate(final ConnectionState connectionState, final Throwable t, @Nullable final ObjectId serviceId) {
+        if (!isClosed()) {
+            if (t instanceof MongoSocketException
+                    && (!(t instanceof MongoSocketReadTimeoutException) || connectionState == BEFORE_HANDSHAKE)) {
+                if (serviceId != null) {
+                    connectionPool.invalidate(serviceId);
+                }
+            } else if (t instanceof MongoNotPrimaryException || t instanceof MongoNodeIsRecoveringException) {
+                if (SHUTDOWN_CODES.contains(((MongoCommandException) t).getErrorCode())) {
+                    if (serviceId != null) {
+                        connectionPool.invalidate(serviceId);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        if (!closed.getAndSet(true)) {
+           serverListener.serverClosed(new ServerClosedEvent(serverId));
+        }
+    }
+
+    @Override
+    public boolean isClosed() {
+        return closed.get();
+    }
+
+    @Override
+    public void connect() {
+        // no op
+    }
+
+    @Override
+    public Connection getConnection() {
+        isTrue("open", !isClosed());
+        return connectionFactory.create(connectionPool.get(), new LoadBalancedServerProtocolExecutor(),
+                ClusterConnectionMode.LOAD_BALANCED);
+    }
+
+    @Override
+    public void getConnectionAsync(final SingleResultCallback<AsyncConnection> callback) {
+        throw new UnsupportedOperationException();
+    }
+
+    private class LoadBalancedServerProtocolExecutor implements ProtocolExecutor {
+        @SuppressWarnings("unchecked")
+        @Override
+        public <T> T execute(final CommandProtocol<T> protocol, final InternalConnection connection, final SessionContext sessionContext) {
+            try {
+                protocol.sessionContext(new ClusterClockAdvancingSessionContext(sessionContext, clusterClock));
+                return protocol.execute(connection);
+            } catch (MongoWriteConcernWithResponseException e) {
```

Comment:
I don't actually know what this is for.  It was copied verbatim from DefaultServer

Reply:
Its to do with retryable writes. Commands run with a write concern can fail due to write concern issues but the command response is also required as part of retryable writes spec.